### ISSUE TITLE
F# build of F* rejects 0ul as out of range

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#1053]
It seems that @nikswamy's recent code for checking constant bounds (https://github.com/FStarLang/FStar/commit/a8c1fbc83f240ca8c4b3ea5935aeebd35901619f#diff-0b9e97904ace9e5e55b6cdcd8e842211R642) does not work with the F# build. This appeared in a larger context (https://github.com/FStarLang/FStar/issues/1023#issuecomment-303512663), but it can be simplified to just this:
```
let foo = 0ul
```
The error message is:
```
(Error) 0 is not in the expected range for FStar.UInt32
```
This works just fine with the OCaml build of F*.

**Comment by catalin-hritcu** [comment]
Ouch, this seems caused by integers in the F# build being bounded, and `pow2 32` returning `-1`. Maybe time to fix this blatant unsoundness biting us back? :)

**Comment by nikswamy** [comment]
Sigh. Yes, the F# build of F* is not sound. We've known this for a long time, since we do not use bignums for representing mathematical integers in F#. As such, the F# build is mainly used for bootstrapping currently. Fixing it to use .net bignums should be possible, but it's tedious work. We should do it at some point.

**Comment by catalin-hritcu** [comment]
Removed the duplicate flag since there might be place for short term mitigations targeted only at this particular issue (while still punting on #1054). Personally, I would very much like if could have the F# version run the regular all regressions as well, so that we're forced to keep it up to date. And didn't seem too far from this (#1023), until this came along at least.

**Comment by A-Manning** [comment]
This is fixed in #1283. 

These overflow errors are quite annoying to debug when they pop up, perhaps the F# build should be compiled with the `--checked` flag?

**Comment by A-Manning** [comment]
Can this be closed now?
